src/storage/src/hummock/iterator/concat_delete_range_iterator.rs

```diff
@@ -0,0 +1,279 @@
+                        iter.next_extended_user_key()
+                            .left_user_key
+                            .eq(&FullKey::decode(
+                                &self.sstables[self.idx + 1].key_range.as_ref().unwrap().left,
+                            )
+                            .user_key);
+                    iter.next().await?;
+                    if !iter.is_valid() && last_key_in_sst_start {
+                        self.seek_idx(self.idx + 1, None).await?;
+                        let next_range = self.next_extended_user_key();
+                        debug_assert!(self.is_valid());
+                        if next_range.is_exclude_left_key == exclusive_range_start
+                            && next_range.left_user_key.eq(&FullKey::decode(
+                                &self.sstables[self.idx].key_range.as_ref().unwrap().left,
+                            )
+                            .user_key)
+                        {
+                            self.current.as_mut().unwrap().next().await?;
+                        }
+                        return Ok(());
+                    }
+                } else {
+                    iter.next().await?;
+                }
+                let mut idx = self.idx;
+                while idx + 1 < self.sstables.len() && !self.is_valid() {
+                    self.seek_idx(idx + 1, None).await?;
+                    idx += 1;
+                }
+            }
+        }
+        Ok(())
+    }
+
+    /// Seeks to a table, and then seeks to the key if `seek_key` is given.
+    async fn seek_idx(
+        &mut self,
+        idx: usize,
+        seek_key: Option<UserKey<&[u8]>>,
+    ) -> HummockResult<()> {
+        self.current.take();
+        if idx < self.sstables.len() {
+            if self.sstables[idx].range_tombstone_count == 0 {
+                return Ok(());
+            }
+            let table = self
+                .sstable_store
+                .sstable(&self.sstables[idx], &mut self.stats)
+                .await?;
+            let mut sstable_iter = SstableDeleteRangeIterator::new(table);
+
+            if let Some(key) = seek_key {
+                sstable_iter.seek(key).await?;
+            } else {
+                sstable_iter.rewind().await?;
+            }
+            self.current = Some(sstable_iter);
+            self.idx = idx;
+        }
+        Ok(())
+    }
+}
+
+impl DeleteRangeIterator for ConcatDeleteRangeIterator {
+    type NextFuture<'a> = impl Future<Output = HummockResult<()>> + 'a;
+    type RewindFuture<'a> = impl Future<Output = HummockResult<()>> + 'a;
+    type SeekFuture<'a> = impl Future<Output = HummockResult<()>> + 'a;
+
+    fn next_extended_user_key(&self) -> PointRange<&[u8]> {
+        self.current.as_ref().unwrap().next_extended_user_key()
+    }
+
+    fn current_epoch(&self) -> HummockEpoch {
+        self.current.as_ref().unwrap().current_epoch()
+    }
+
+    fn next_range_epoch(&self) -> HummockEpoch {
```

Comment:
this may return a false `HummockEpoch::MAX` when `next()` is the special situation